### ISSUE TITLE
Design Proposal: Handling Volumes Attached to Inaccessible Nodes

### ISSUE THREAD

**Issue by saad-ali** [kubernetes/kubernetes#15524]
*Update: detailed design doc in #20262*

# Problem

Volumes remain attached to a node when kubelet on the node is unexpectedly terminated (because, for example, the node was restarted). When a node becomes inaccessible, the scheduler, schedules the pods from the downed node to another node. However, certain volumes can only be attached to one node at a time (for example GCE PDs may only be attached to one instance in RW mode). Pods that require a volume that can only be attached to a single node will be unable to start once rescheduled to a new node, because their volume can not be attached to the new node since it was never detached from the old node.

See issue https://github.com/kubernetes/kubernetes/issues/14642

# Potential Solutions
1. Node detaches from other node
  * If a node detects that it is unable to attach a device because it is attached to another node, the node should detach it from the other node, and retry attaching it.
  * Cons: Could lead to two nodes fighting each other for a volume.
2. Garbage collection on node
  * Cons: Can’t rely on the node to to do cleanup: for example garbage collection can’t happen if the plug is pulled on a node.
3. Node Fencing
  * The logic on the scheduler that currently determines if a node is up or down and reschedules pods, should also trigger resource reclamation from downed nodes (detaching disks as well as potentially other clean up).
  * Cons: Decoupling attachment and detachment leads to inherent race conditions, as long as the node itself can decide to attach a volume to itself.
4. Volume Attachment/Detachment Controller
  * Move the volume attachment/detachment logic from kubelet to a separate off-node controller (maybe running on master or as a pod).

# Proposal

I think the “Volume Attachment/Detachment Controller” would be the most robust design.

A node should not be responsible for determining whether to attach or detach a volume to itself. Instead the responsibility should be moved (out of kubelet) to a separate off-node “Volume Attachment/Detachment Controller”.

This controller would keep track of:
* Which pods require which volumes.
* Which nodes these pods have been scheduled to.
* Which volumes have been successfully attached to which nodes.

The controller would be responsible for attaching a volume to a node if:
* A pod with a volume is scheduled to a node and the volume is not already attached to the node (no existing pod on the node requires the volume).
* The volumes policy permits attachment to multiple nodes, or the volumes policy permits attachment only to a single node and there are no other nodes that the volume is already attached to.

The controller would be responsible for detaching a volume from a node if:
* There are no pods running on the node using that volume.

Some volume types may not have a concept of attaching to a node (e.g. NFS volume), and for these volume types the attach and detach methods should be no-ops.

Another benefit of this approach is that it allows the controller to ensure that attach and detach do not occur simultaneously for a particular volume (avoiding lots of horrible race conditions).

## Open questions:
* Mount/unmount code must still exist on the node. How does the controller signal to the node that a volume is attached along with sufficient information (mount dir, etc.) for the node to mount it?
  * Could the mounting logic be dumb and just constantly poll a location on a disk? Would it work for all existing and future volumes?
  * Could the attach/detach logic update the PVC/PV object with some state indicating the nodes it is attached to?
* How do you make this controller plugable?
  * As we proposed for the dynamic volume provisioning controller.
    * For v1 all plugins are compiled in (as is the case for existing volume “plugins”).
    * For v2 anyone can implement their own attachment/detachment controller by creating a binary/pod/script/etc. that:
      * Monitoring API server for pod, node, and PVC/PV objects.
      * Writing updates to these objects

CC @thockin @markturansky

**Comment by markturansky** [comment]
Two additional features the suggested controller can provider:

1. Enforce resource constraints - https://github.com/kubernetes/kubernetes/issues/7835
2. Secure cloud credentials to master - https://github.com/kubernetes/kubernetes/issues/12399

**Comment by markturansky** [comment]
This problem effects EBS as well: http://www.fpaste.org/279970/05660144/

**Comment by saad-ali** [comment]
Yep, and the proposed solution would fix the issue for all volumes.

**Comment by markturansky** [comment]
@kubernetes/rh-storage

**Comment by rootfs** [comment]
+1 the proposed solution. 

More broadly, since the proposed controller essentially determines volume affinity,  we should not just limit ourselves to detachment/attachment. Other policies are also manageable by this controller. For example, since networking storage consume the same underlying network fabric, it is ideal for the volume affinity controller to spread out Pods to different hosts/subnet so Pods attaching to different volumes are not competing on host network bandwidth.

**Comment by saad-ali** [comment]
@rootfs That would be an interesting dimension for scheduling. However, this proposed attachment/detachment controller is not be responsible for scheduling pods. It is only responsible for making sure that volumes are attached to the correct node once pods are already scheduled and detached upon completion.

**Comment by chrislovecnm** [comment]
@saad-ali do we have a final design on this?

**Comment by ddysher** [comment]
@saad-ali what's the status on this?  We've been bitten by this twice :(

> Mount/unmount code must still exist on the node. How does the controller signal to the node that a volume is attached along with sufficient information (mount dir, etc.) for the node to mount it?

We've previously tried to scan /sys/block for upcoming volume, but I'm not sure it will work for future volumes.  Kubelet can also poll cloudprovider for volume information, provided that cloudprovider will have mount dir of the volume.  Actually, is it possible to have the controller just detecting incorrect volume, and have kubelet do the attachment?
> Could the mounting logic be dumb and just constantly poll a location on a disk? Would it work for all existing and future volumes?

Will this work for Pod that doesn't use PVC? i.e. use a pre-allocated volume.  I didn't follow PVC/PV but guess it's not encouraged as of now? @markturansky
> Could the attach/detach logic update the PVC/PV object with some state indicating the nodes it is attached to?

**Comment by markturansky** [comment]
@ddysher Persistent Volume vs. regular Volume should be irrelevant when Attaching/Detaching through a single controller.  

Red Hat has a requirement for this feature, too.  We're happy to do the work but we'll need to go over technical design with you, @saad-ali.

**Comment by saad-ali** [comment]
> @saad-ali do we have a final design on this?

We're just wrapping up the Kubernetes v1.1 release. I'm starting to work on a detailed design for this proposal.

> We've been bitten by this twice :(

Sorry to hear that :( We'll try to get this fixed ASAP

> We've previously tried to scan /sys/block for upcoming volume, but I'm not sure it will work for future volumes. Kubelet can also poll cloudprovider for volume information, provided that cloudprovider will have mount dir of the volume. Actually, is it possible to have the controller just detecting incorrect volume, and have kubelet do the attachment?

We have two options here. Either the controller provides a signal to the kubelet indicating that a volume is attached or it provides no such signal. If it does, then the kubelet will wait until that signal before trying to mount. If it doesn't, then it is up to each individual volume plugin mount logic to try and figure out if the volume has been attached or not (it could do this via `/sys/block` scan as you mentioned, or by polling `/dev/disk/by-id/` as GCE PD currently does, or some other mechanism), and to wait and retry if it isn't. I'm currently leaning towards the controller providing a signal, it'll simplify the mount logic for most plugins.

> Will this work for Pod that doesn't use PVC? i.e. use a pre-allocated volume. I didn't follow PVC/PV but guess it's not encouraged as of now? @markturansky

Like @markturansky mentioned as far as attach/detach logic is concerned it doesn't matter if the volume is a PVC or PV/pre-allocated or dynamically created.

> Red Hat has a requirement for this feature, too. We're happy to do the work but we'll need to go over technical design with you, @saad-ali.

Thanks @markturansky. I'm working on a detailed design proposal. I'll make sure to get your/RedHat's input on it. We should be able to handle the work (coding) on our end, but we'll be sure to let you know if that changes.

**Comment by ddysher** [comment]
@saad-ali Thanks for your work!

Signal sounds like a solution. I'm a little blank on how this work in kubernetes, and how we make sure the signal is delivered robustly.  Look forward to see the detailed proposal!
> We have two options here. Either the controller provides a signal to the kubelet indicating that a volume is attached or it provides no such signal. If it does, then the kubelet will wait until that signal before trying to mount. If it doesn't, then it is up to each individual volume plugin mount logic to try and figure out if the volume has been attached or not (it could do this via /sys/block scan as you mentioned, or by polling /dev/disk/by-id/ as GCE PD currently does, or some other mechanism), and to wait and retry if it isn't. I'm currently leaning towards the controller providing a signal, it'll simplify the mount logic for most plugins.

**Comment by chrislovecnm** [comment]
@saad-ali let me know when we will have a timeline on this ... bosses are asking :smile:

**Comment by pmorie** [comment]
@kubernetes/rh-cluster-infra

**Comment by pmorie** [comment]
@saad-ali Do you mean this to apply to persistent volumes only?  Ie, if I use a raw aws ebs volume, do I get this treatment?

**Comment by pmorie** [comment]
@saad-ali If we only mean this to apply to PVs, the mount controller can annotate the PV and the kubelet can watch those annotations.  That might be better than adding an API to the kubelet.

**Comment by markturansky** [comment]
@pmorie the requirement to attach volumes exists for regular volumes and PV usage.  The controller will need to do both.  Regular volumes in pods are probably easier.  No need to look up what backs the persistent-volume plugin.

**Comment by pmorie** [comment]
@markturansky What I'm asking is whether volumes not wrapped in a PV also get this treatment.

@saad-ali Also, which volume types get this treatment?  System-generated volumes like secrets, emptyDir, and downward API don't need it (afaict).  How do we distinguish the two classes of volumes?  Is it an attribute, like the ones discussed in #16614 ?  cc @vishh

**Comment by markturansky** [comment]
@pmorie I understand.  I'm saying this should also apply to volumes not wrapped in a PV.

**Comment by markturansky** [comment]
@pmorie Clarification, this applies, I believe, to volumes that require attaching and that applies regardless of whether that volume is within a PV or not.

**Comment by pmorie** [comment]
@markturansky @saad-ali I'm having bad luck w/ GH on the road today...

Seems like this implies that we would annotate both pods and PVs -- let's have all the mechanics spelled out in the design proposal.

I am pretty stoked about this, looking forward to seeing more :)

**Comment by swagiaal** [comment]
> I think the “Volume Attachment/Detachment Controller” would be the most robust design.

+1

> Could the mounting logic be dumb and just constantly poll a location on a disk? Would it work for all existing and future volumes?

That what all volume plugins do currently. I think the first implementation of this should just rely on that. We can use inotify to avoid polling

**Comment by swagiaal** [comment]
> That what all volume plugins do currently. I think the first implementation of this should just rely on that. We can use inotify to avoid polling

On second though I think that polling/inotify will always be needed. Plugins do this currently despite the api returning a confirmation of the attach meaning that there is a lag between confirmation from the api and the actual device appearing in `/dev/disk/by-id`. The Attach/Detach wont have any extra knowledge to eliminate this lag

**Comment by thockin** [comment]
> How does the controller signal to the node that a volume is attached

The nodes have to have the responsibility for making device nodes for attached
devices, and for making those nodes available to kubelet in a way that kubelet
can identify the device - this is what udev is for, right?

> Could the attach/detach logic update the PVC/PV

Status on the PV seems appropriate

> How do you make this controller plugable?

Medium term we are moving to a per-cloud-provider storage provisioner, right?
Should that expand to be a per-cloud-provider storage manager - provisioning, 
recycling, attaching, etc?

@rootfs

> the volume affinity controller to spread out Pods to different hosts

This is a much more significant change - it influences scheduling.  Let's not 
throw too much in at one time.


Is it worth sketching pseudocode for the controller here, so we can haggle over semantics while they are cheap to change?

**Comment by markturansky** [comment]
This proposal will need to account for the race between the detaching controller and the controller that deletes the PV https://github.com/kubernetes/kubernetes/issues/18832

**Comment by chrislovecnm** [comment]
@saad-ali how is this progressing??

**Comment by childsb** [comment]
OpenShift Online is blocked on an a security hole related to this re-design : https://github.com/kubernetes/kubernetes/issues/12399

How can we help move this along?  Is there any chance for 1.2?

**Comment by saad-ali** [comment]
I have a preliminary detailed design doc to move attach/detach logic to a controller. I haven't shared it publicly yet because I want to make sure it's in alignment with whatever we decide for https://github.com/kubernetes/kubernetes/issues/18333. We're planning on closing on https://github.com/kubernetes/kubernetes/issues/18333 next week (Storage SIG meeting 11 AM PST Jan 5). Once that happens, I'm planning on publishing the design doc and starting coding soon after.

**Comment by markturansky** [comment]
Added benefit of centralized controller (just because this came up in an internal email discussion):

A controller watching pods can immediately start detaching a volume when its no longer needed.  Contrast this with the Kubelet's sync loop that has to discover the orphaned volume.  Latency for the next pod to launch is reduced because detaching happens sooner and can be attached to the next node more quickly.

**Comment by saad-ali** [comment]
Here is the detailed design doc for the design we are currently implementing: https://github.com/kubernetes/kubernetes/issues/20262. This will not make it into v1.2 but should be in v1.3.

**Comment by thockin** [comment]
Superceded by https://github.com/kubernetes/kubernetes/issues/20262
